### PR TITLE
oop-user_defined_classes-hard-q4

### PR DESCRIPTION
**Pull request by SamamthaH**
## PR Checklist
- [x]  Are helper methods really needed here?
    > Opt for hardcoding input/output into the input provider when possible
- [x]  Are the assertion fail messages good?
    - [x]  Logical (does it make sense?)
    - [x]  Readable (can a student understand what’s wrong?)
    - [x]  Not too feed-y (does it give away the answer?)
    - [x]  Grammatically sound (is the grammar and punctuation in the sentence proper?)
    - [x]  Specific (does the fail message relate to the question?)
- [x]  Is the code formatted well?
    > Run `CTRL` + `ALT` + `L` in IntelliJ
- [x]  EOF newlines
- [x]  Do the comments for Parsons follow the correct convention?
- [x]  Is the question being tested sufficiently?
    - [x]  Are there representative cases to test the program?
    - [x]  What about edge cases? (empty strings, negatives, infinity, off-by-one errors, etc)
        > If the question doesn’t specify what to do, mark it as a bad question
- [x]  Are the test method names good?
    > Test method names show up in a “What went well” section on the website, so “mathTest” will become “What went well: math test”, which doesn’t make much sense. A better name in this example would be “calculatesTaxCorrectly” so it shows up as “What went well: calculates tax correctly”

### CODE REVIEW

File: src/oop/user_defined_classes/hard/q4/MainTest.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+                String.format(incorrectModifierMessage, intAttributeName1, testClassName));
+    }
+
+    @Test
+    public void laundromatClassHasRequiredConstructor() {
+        String missingConstructorMessage = "Your " + testClassName + " class is missing a required constructor.";
+        Class<?>[] arguments = {
+                int.class
+        };
+        assertTrue(classInstance.hasConstructor(arguments), missingConstructorMessage);
```
Comment (KaiKhademi) [inline src/oop/user_defined_classes/hard/q4/MainTest.java:48]: Also check the constructor modifier

File: src/oop/user_defined_classes/hard/q4/MainTest.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+    @Test
+    public void laundromatClassHasCorrectAttributes() {
+        String missingAttributeMessage =
+                "The %s attribute could not be found in your %s class. Please make sure you have added it, it is spelled correctly, and is of the correct type";
+        String incorrectModifierMessage = "Your %s attribute in the %s class does not have the correct visibility modifier.";
+        assertTrue(classInstance.hasField(intAttributeName1, int.class),
+                String.format(missingAttributeMessage, intAttributeName1, testClassName));
+        assertTrue(classInstance.hasModifier(intAttributeName1, "private"),
+                String.format(incorrectModifierMessage, intAttributeName1, testClassName));
```
Comment (KaiKhademi) [inline src/oop/user_defined_classes/hard/q4/MainTest.java:39]: Mention which attribute has a problem in the error messages.

File: src/oop/user_defined_classes/hard/q4/MainTest.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    }
+
+    @ParameterizedTest
+    @MethodSource("methodInputProvider")
+    public void correctDoLaundryMethod(int attribute1) throws Throwable {
+        Object[][] arguments = {
+                {attribute1, int.class}
+        };
+        Object instance = classInstance.createInstance(arguments);
+        assertTrue(classInstance.hasMethod(methodName, null),
```
Comment (KaiKhademi) [inline src/oop/user_defined_classes/hard/q4/MainTest.java:79]: Check the return type and modifier here.

File: src/oop/user_defined_classes/hard/q4/QuestionText.md [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+down the street? Write a Java class called `Laundromat` to keep track of dirty items and let you know when they are 
+clean. See the description below for the requirements.
+
+### Laundromat Class:
+
+Your `Laundromat` class should have the following items:
+
+- Attributes:
+    - An `int` called **gymClothes** which should not be accessible from another class
+
```
Comment (KaiKhademi) [inline src/oop/user_defined_classes/hard/q4/QuestionText.md:21]: Mention the visibility modifier for the constructors and methods as well

**Review by KaiKhademi — CHANGES_REQUESTED** [review]

File: src/oop/user_defined_classes/hard/q4/QuestionText.md (hunk 4 above)
Comment (SamamthaH) [reply src/oop/user_defined_classes/hard/q4/QuestionText.md:21]: Wouldn't a private constructor for an object be redundant? If no other class can access it? Also, I was using the "which another class can use" in the method description to signify that the method should be accessible from another class

File: src/oop/user_defined_classes/hard/q4/MainTest.java [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+    }
+
+    @Test
+    public void laundromatClassHasRequiredConstructor() {
+        Class<?>[] arguments = {
+                int.class
+        };
+        assertTrue(classInstance.hasConstructor(arguments),
+                String.format("Your %s class is missing a required constructor.", testClassName));
+        assertTrue(classInstance.hasConstructor(arguments, new String[]{"public"}),
```
Comment (KaiKhademi) [inline src/oop/user_defined_classes/hard/q4/MainTest.java:49]: Same here as the previous PRs

File: src/oop/user_defined_classes/hard/q4/QuestionText.md (hunk 4 above)
Comment (KaiKhademi) [reply src/oop/user_defined_classes/hard/q4/QuestionText.md:21]: They might now add a modifier for constructor which would be the default modifier and is accessible from other classes in the same package. I believe for these questions we would want them to make their constructors public.

Comment (KaiKhademi) [reply src/oop/user_defined_classes/hard/q4/QuestionText.md:21]: *They might not add

**Review by KaiKhademi — CHANGES_REQUESTED** [review]

File: src/oop/user_defined_classes/hard/q4/QuestionText.md [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+- Constructors:
+    - A constructor which accepts one int and sets the **gymClothes** attribute.
+    - This constructor should be visible to other classes
+
+- Methods:
+    - A method called `doLaundry` which another class can use to clean the **gymClothes** one by one. It displays a 
+      message to let the user know that each item is being cleaned. When all the items have been cleaned, the method 
+      displays that the laundry is done. **gymClothes** should be set to 0 afterwards. Sample output of this method when
```
Comment (Hedgemon4) [inline src/oop/user_defined_classes/hard/q4/QuestionText.md:29]:
```suggestion
      displays that the laundry is done. **gymClothes** should be set to 0 afterwards. The sample output of this method when 
```

File: src/oop/user_defined_classes/hard/q4/MainTest.java [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+                String.format(incorrectModifierMessage, intAttributeName1, testClassName));
+    }
+
+    @Test
+    public void laundromatClassHasRequiredConstructor() {
+        Class<?>[] arguments = {
+                int.class
+        };
+        assertTrue(classInstance.hasConstructor(arguments),
+                String.format("Your %s class is missing a required constructor.", testClassName));
```
Comment (Hedgemon4) [inline src/oop/user_defined_classes/hard/q4/MainTest.java:48]: Maybe add "Please check that you have added the constructor and it has the correct parameters."

File: src/oop/user_defined_classes/hard/q4/MainTest.java [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+
+    @ParameterizedTest
+    @MethodSource("methodInputProvider")
+    public void correctDoLaundryMethod(int attribute1) throws Throwable {
+        Object[][] arguments = {
+                {attribute1, int.class}
+        };
+        Object instance = classInstance.createInstance(arguments);
+        assertTrue(classInstance.hasMethod(methodName, null, Void.TYPE, new String[]{"public"}),
+                "Your " + testClassName + " class is missing the method " + methodName + " or it does not have the correct return type or visibility modifier.");
```
Comment (Hedgemon4) [inline src/oop/user_defined_classes/hard/q4/MainTest.java:82]: You should test these separately

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

**Review by KaiKhademi — DISMISSED** [review]

File: src/oop/user_defined_classes/hard/q4/QuestionText.md [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+  
+      Cleaning item 2
+  
+      Cleaning item 3
+  
+      Cleaning item 4
+  
+      Cleaning item 5
+  
+      Your laundry is done!
```
Comment (Hedgemon4) [inline src/oop/user_defined_classes/hard/q4/QuestionText.md:42]:
```suggestion
      Cleaning item 1
      Cleaning item 2
      Cleaning item 3
      Cleaning item 4
      Cleaning item 5
      Your laundry is done!
```

Comment (Hedgemon4) [reply src/oop/user_defined_classes/hard/q4/QuestionText.md:42]: From the look of the method, this is the actually output

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

**Review by Hedgemon4 — APPROVED** [review]

**Pull request merged**